src/utilities/codemirror/custom-folder.ts

```diff
@@ -1,44 +1,178 @@
-    return foldCommand(node, state);
+  switch (node.name) {
+    case 'Activate':
+    case 'Load':
+      return foldActivateLoad(node, state);
+    case 'Command':
+      return foldCommand(node, state);
+    case 'GroundBlock':
+    case 'GroundEvent':
+      return foldGround(node, state);
+    case 'Request':
+      return foldRequest(node, state);
+    case 'Metadata':
+      return foldMetadataOrModel(node, state, '@Metadata');
+    case 'Models':
+      return foldMetadataOrModel(node, state, '@Model');
   }
   return null;
 }
 
-function foldCommand(node: SyntaxNode, state: EditorState): { from: number; to: number } | null {
-  const stemNode = node.getChild('Stem');
-  const argsNodes = node.getChildren('Args');
-  const commentNode = node.getChild('LineComment');
-  const metadataNode = node.getChildren('Metadata');
-  const modelNodes = node.getChildren('Models');
-
+/**
+ * Returns the range of the `Command` node, folding the node on the Stem name.
+ * @param commandNode The `Command` node to fold.
+ * @param state The EditorState to get the document from.
+ * @returns The range of the `Command` node, or null if the Stem node is missing.
+ */
+export function foldCommand(commandNode: SyntaxNode, state: EditorState): { from: number; to: number } | null {
+  const stemNode = commandNode.getChild('Stem');
   if (stemNode == null) {
     return null;
   }
 
-  const from = calculateStartAndEnd([stemNode, ...argsNodes, commentNode]).to;
-  const nonCommandTo = calculateStartAndEnd([...metadataNode, ...modelNodes]).to;
-  const text = state.sliceDoc(from, nonCommandTo);
-  // Exclude the last new line so commands remain on different lines
+  // fold the node on the Stem name
+  return foldSteps(commandNode, stemNode, state);
+}
+
+/**
+ * Returns the range of the `GroundBlock` or `GroundEvent` node, folding the node on the GroundName.
+ * @param groundNode The `GroundBlock` or `GroundEvent` node to fold.
+ * @param state The EditorState to get the document from.
+ * @returns The range of the `GroundBlock` or `GroundEvent` node, or null if the GroundName node is missing.
+ */
+export function foldGround(groundNode: SyntaxNode, state: EditorState): { from: number; to: number } | null {
+  // Get the GroundName node
+  const groundNameNode = groundNode.getChild('GroundName');
+  // If the GroundName node is missing, return null
+  if (groundNameNode == null) {
+    return null;
+  }
+
+  // Fold the node on the GroundName
+  return foldSteps(groundNode, groundNameNode, state);
+}
+
+/**
+ * Returns the range of the `Activate` or `Load` node, folding the node on the `SequenceName`.
+ * @param activateNode The `Activate` or `Load` node to fold.
+ * @param state The EditorState to get the document from.
+ * @returns The range of the `Activate` or `Load` node, or null if the `SequenceName` node is missing.
+ */
+export function foldActivateLoad(activateNode: SyntaxNode, state: EditorState): { from: number; to: number } | null {
```

Comment:
Can we collapse these 3 functions into one where we pass the node and the node name because they all are functionally the same?

Reply:
Commands and Ground Commands are slightly different so it isn't as simple as merging everything together into one function call. For commands ,you have to step a node deeper and pull out the 'stem' node' before processing.

Reply:
Doesn't:

```
export function foldGeneric(node: SyntaxNode, name: string, state: EditorState): { from: number, to: number } | null {
  const nameNode = node.getChild(name);

  if (nameNode === null) {
    return null;
  }

  return foldSteps(node, nameNode, state);
}
```

Do the same thing as these 3 functions? It's not a huge deal I'm just curious.

Reply:
ahhh, I see, sorry my brain was fried by the end of the day on Friday. I removed the 3 functions and modified the `foldStep` function to be called by the switch statement.